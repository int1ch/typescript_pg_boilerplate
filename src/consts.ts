export const SMS_NOTIFICATION = "sms_notification";
export const EMAIL_NOTIFICATION = "email_notification";

export const UserConsentTypeVariants = [
  SMS_NOTIFICATION,
  EMAIL_NOTIFICATION,
] as const;
export type UserConsentType = typeof UserConsentTypeVariants[number];
