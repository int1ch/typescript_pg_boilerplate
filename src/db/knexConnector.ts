import { knex } from "knex";
import { snakecase, camelcase } from "stringcase";
import * as config from "../config";
import logger from "../logger";

export async function testConnection() {
  await connection.raw("SELECT now()");
  return connection;
}

const DUMP = false;
const DEBUG = false;

export function openConnection() {
  //ups, dev may run in testing / development
  //other envs expected to use only PG_PASSWORD,
  const password = config.IS_TESTING
    ? config.PG_TESTING_PASSWORD
    : config.PG_PASSWORD;

  const connection = knex({
    client: "pg",
    connection: {
      host: config.PG_HOST,
      port: config.PG_PORT,
      user: config.PG_USERNAME,
      password: password,
      database: config.PG_DATABASE,
    },
    acquireConnectionTimeout: 3000,
    pool: {
      min: 2,
      max: 2, //FIXME
    },
    wrapIdentifier: snakeWrapper,
    postProcessResponse: postConverter,
    debug: DEBUG,
  });
  return connection;
}

function snakeWrapper(
  value: string,
  origImpl: (value: string) => string,
  queryContext: any
): string {
  if (DUMP) logger.log("snake wrapper:", value, origImpl, queryContext);

  return origImpl(snakecase(value));
}
function postConverter(value: any, queryContext: any, path?: string): any {
  if (isRaw(value, queryContext, path)) return value;
  if (DUMP) logger.log("Query POST Converter:", value);
  if (Array.isArray(value)) {
    return value.map((item) => postConverter(item, queryContext));
  }
  if (value._parsers && value._types) {
    const output: any = {};
    Object.assign(output, value);
    output.rows = postConverter(value.rows, queryContext);
    return output;
  }

  const output: any = {};
  path ||= "";
  for (const key of Object.keys(value)) {
    const newKey = camelcase(key);
    output[newKey] = value[key];
    /*postConverter(
      value[key],
      queryContext,
      (path += "." + key)
    );*/
  }

  return output;
}

function isRaw(value: any, queryContext: any, path?: string) {
  if (typeof value !== "object" || value === null) return true;
  if (value instanceof Date) return true;
  return false;

  //if (path === "root") return false;
  //if (typeof recursive !== "function") return true;
  //return !recursive(value, path, queryContext);
}

export async function closeConnection() {
  connection.destroy();
  //logger.log("unsupported?");
  //connection.close
}
const connection = openConnection();
export default connection;
