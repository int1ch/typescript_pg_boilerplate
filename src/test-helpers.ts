import { userInfo } from "os";
import dbUser, { DbUser } from "./db/User";

export function fixUuidIn(body: any, field: string) {
  try {
    body[field] = "uuid0000-fixe-df04-test-reason6922cd";
  } catch (e) {}
}

export function fixEmailIn(body: any, field: string) {
  try {
    body[field] = "fixed@email.ru";
  } catch (e) {}
}

function randInt(max = 100000) {
  Math.ceil(Math.random() * 10000);
}

export class UserCreator {
  protected dbUser: DbUser;
  protected createdIds: string[];
  protected n = 0;
  constructor() {
    this.dbUser = dbUser;
    this.createdIds = [];
  }
  async create(options?: { email?: string }) {
    const rand = randInt();
    const email = options?.email || `cr-${this.n++}-${rand}@test-cr.com`;

    const userId = await this.dbUser.create({ email: email });
    this.createdIds.push(userId);
    return userId;
  }
  async cleanUp() {
    for (const userId of this.createdIds) {
      await this.dbUser.delete(userId);
    }
  }
}

export const userCreatorHelper = new UserCreator();
