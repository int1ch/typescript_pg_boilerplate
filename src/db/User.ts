import { RandomUUIDOptions } from "crypto";
import { v4 as UUID } from "uuid";

import { Knex } from "knex";
import dbConnection from "./knexConnector";
import { UserConsentType } from "../consts";

export type UserId = string;

interface UserOptions {
  email: string;
}
export interface IUser {
  userId: UserId;
  email: string;
}

export interface UserConsent {
  userId: UserId;
  type: UserConsentType;
  enabled: boolean;
}

export interface UserConsentHistoryEvent extends UserConsent {
  updated_at: Date;
}

export class EmailAlreadyUsedError extends Error {}

export class DbUser {
  constructor(protected knex: Knex = dbConnection) {}

  /*
   * @throws {EmailAlreadyUsedError}
   */
  protected interceptAlreadyUsedEmail(e: any, email?: string) {
    //console.error({ error: e }, "INTERCEPTOR");
    if (String(e.message).includes("users_email_unique")) {
      throw new EmailAlreadyUsedError(
        `Email ${email} already used by some one`
      );
    }
  }

  /*
   *
   * @throws {EmailAlreadyUsedError}
   */
  async create(user: UserOptions) {
    const toCreate = { email: user.email, userId: "" };
    const db = this.knex;
    let createdUserId: string | undefined;
    let tryCounter = 0;

    while (!createdUserId) {
      if (tryCounter > 2) {
        throw new Error("Cannot create user; db error");
      }
      let uuid = UUID();
      toCreate.userId = uuid;
      try {
        const created = await db<IUser>("users")
          .insert(toCreate)
          .returning("userId");
        createdUserId = created[0];
      } catch (e: any) {
        this.interceptAlreadyUsedEmail(e, toCreate.email);
        throw e;
      }
      tryCounter++;
    }
    return createdUserId;
  }
  async update(user: IUser) {
    const userId = user.userId;
    const toUpdate = { email: user.email };
    const db = this.knex;
    try {
      await db<IUser>("users").where({ userId }).update(toUpdate);
    } catch (e: any) {
      this.interceptAlreadyUsedEmail(e, toUpdate.email);
      throw e;
    }
  }
  async delete(userId: UserId) {
    const db = this.knex;
    await db("users").where({ userId }).delete();
  }
  async get(userId: UserId) {
    //logger.debug("Fetcheng by userID:", userId);
    const user = await this.knex<IUser>("users").where({ userId }).first();
    return user;
  }

  async getUserConsents(userId: UserId): Promise<UserConsent[]> {
    const db = this.knex;
    const consents = await db<UserConsent>("user_consents").where({ userId });
    return consents.sort((a, b) => a.type.localeCompare(b.type));
  }
  async getUserConsentsHistory(userId: UserId, offset = 0, limit = 20) {
    const db = this.knex;
    const consents = await db<UserConsentHistoryEvent>("user_consents_history")
      .where({ userId })
      .orderBy("updated_at", "desc")
      .limit(limit)
      .offset(offset);
    return consents;
  }
  async setUserConsent(consent: UserConsent) {
    const db = this.knex;
    await db.transaction(async (trx) => {
      await trx("user_consents_history").insert(consent);
      await trx("user_consents")
        .insert(consent)
        .onConflict(["user_id", "type"])
        .merge({ enabled: consent.enabled });
    });
    //auto commit expected
  }
}

export default new DbUser();
