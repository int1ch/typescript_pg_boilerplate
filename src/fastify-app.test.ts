import { EMAIL_NOTIFICATION, SMS_NOTIFICATION } from "./consts";
import { closeConnection } from "./db/knexConnector";
import dbUser, { DbUser, UserConsent } from "./db/User";
import appBuilder from "./fastify-app";
import { fixEmailIn, fixUuidIn, userCreatorHelper } from "./test-helpers";

const app = appBuilder({});

const rand = Math.ceil(Math.random() * 10000);
const emailToCreate = `fastify-app${rand}@test.com`;
const emailToCreateForAll = `fastify-gar${rand}@test.com`;
const emailForConsents = `fastify2-app${rand}@test.com`;
let createdUserId: string | undefined;
let createdForAllUserId: string | undefined;

const createdUsersIds: string[] = [];
const fakeUserUd = "d290f1ee-6c54-4b01-90e6-d701748f0000";

beforeAll(async () => {
  const userId = await dbUser.create({ email: emailToCreateForAll });
  createdForAllUserId = userId;
  createdUsersIds.push(userId);
});

afterAll(async () => {
  await cleanUp();
  await userCreatorHelper.cleanUp();
  await closeConnection();
});

async function cleanUp() {
  for (const userId of createdUsersIds) {
    await dbUser.delete(userId);
  }
}

describe("HEALTH", () => {
  it("status ok", async () => {
    const response = await app.inject({
      method: "GET",
      url: "/status",
    });
    expect(response.statusCode).toBe(200);
    expect(response.json().ok).toBe(true);
  });
});

describe("POST USERS / create", () => {
  it("BAD EMAIL", async () => {
    const response = await app.inject({
      method: "POST",
      url: "/users/",
      payload: {
        email: "thisIs@email",
      },
    });
    expect(response.statusCode).toBe(422);
    expect(response.json()).toMatchSnapshot();
  });
  it("NO EMAIL", async () => {
    const response = await app.inject({
      method: "POST",
      url: "/users/",
      payload: {},
    });
    expect(response.statusCode).toBe(422);
    expect(response.json()).toMatchSnapshot();
  });

  it("CREATED BY EMAIL", async () => {
    const response = await app.inject({
      method: "POST",
      url: "/users/",
      payload: {
        email: emailToCreate,
      },
    });
    expect(response.statusCode).toBe(200);
    const jsonResponse = response.json();
    expect(jsonResponse).toBeTruthy();
    createdUserId = jsonResponse.id;

    expect(createdUserId).toBeTruthy();
    createdUsersIds.push(createdUserId!);
    fixUuidIn(jsonResponse, "id");
    fixEmailIn(jsonResponse, "email");
    expect(jsonResponse).toMatchSnapshot();
  });
});
describe("GET USERS", () => {
  it("BAD ID", async () => {
    const response = await app.inject({
      method: "GET",
      url: "/users/",
    });
    expect(response.statusCode).toBe(422);
    //expect(response.json()).toMatchSnapshot();
  });
  it("BAD ID", async () => {
    const response = await app.inject({
      method: "GET",
      url: "/users/1",
    });
    expect(response.statusCode).toBe(422);
    expect(response.json()).toMatchSnapshot();
  });
  it("404 by UUID", async () => {
    const response = await app.inject({
      method: "GET",
      url: "/users/d290f1ee-6c54-4b01-90e6-d701748f0851",
    });
    expect(response.statusCode).toBe(404);
    expect(response.json()).toMatchSnapshot();
  });
  it("200 by UUID", async () => {
    const response = await app.inject({
      method: "GET",
      url: `/users/${createdForAllUserId}`,
    });
    expect(response.statusCode).toBe(200);
    const responseJson = response.json();
    fixUuidIn(responseJson, "id");
    fixEmailIn(responseJson, "email");
    expect(responseJson).toMatchSnapshot();

    expect(responseJson.id).toBeTruthy();
    expect(responseJson.email).toBeTruthy();
    expect(responseJson.consents).toHaveLength(0);
  });
});

describe("POST USER CONSENT EVENT", () => {
  it("user created", () => {
    expect(createdUserId).toBeTruthy();
  });

  it.each([
    ["sms_notification", true],
    ["email_notification", true],
    ["sms_validation", false],
  ])("EVENT VALIDATION consent %s isValid: %s", async (consentId, isValid) => {
    const response = await app.inject({
      method: "POST",
      url: `/events`,
      payload: {
        user: {
          id: createdForAllUserId,
        },
        consents: [
          {
            id: consentId,
            enabled: true,
          },
        ],
      },
    });
    if (isValid) {
      expect(response.statusCode).toEqual(200);
    } else {
      expect(response.statusCode).toEqual(422);
    }
  });
  it("POST USER CONSENT EVENT check insert insert", async () => {
    const userId = await userCreatorHelper.create();
    interface event {
      id: string;
      enabled: boolean;
    }
    async function postEvents(userId: string, events: event | event[]) {
      if (!Array.isArray(events)) {
        events = [events];
      }
      return await app.inject({
        method: "POST",
        url: `/events`,
        payload: {
          user: {
            id: userId,
          },
          consents: events,
        },
      });
    }
    const response1 = await postEvents(userId, {
      id: SMS_NOTIFICATION,
      enabled: true,
    });
    expect(response1.statusCode).toBe(200);
    let loggedEvents = await dbUser.getUserConsentsHistory(userId);
    expect(loggedEvents).toHaveLength(1);
    const response2 = await postEvents(userId, {
      id: EMAIL_NOTIFICATION,
      enabled: false,
    });
    expect(response2.statusCode).toBe(200);
    loggedEvents = await dbUser.getUserConsentsHistory(userId);
    expect(loggedEvents).toHaveLength(2);
    let consents = await dbUser.getUserConsents(userId);
    fixUuidInConents(consents);
    expect(consents).toMatchSnapshot();

    const response3 = await postEvents(userId, [
      { id: SMS_NOTIFICATION, enabled: false },
      {
        id: EMAIL_NOTIFICATION,
        enabled: true,
      },
    ]);
    expect(response3.statusCode).toBe(200);
    consents = await dbUser.getUserConsents(userId);
    fixUuidInConents(consents);
    expect(consents).toMatchSnapshot();
    expect(consents[0].enabled).toBe(true); //EMAIL
    expect(consents[1].enabled).toBe(false); //SMS
    loggedEvents = await dbUser.getUserConsentsHistory(userId);
    expect(loggedEvents).toHaveLength(4);
  });
});

describe("DELETE USERS", () => {
  const emailForDelete = `fastify-del${rand}@test.com`;

  async function createUser() {
    const userId = await dbUser.create({ email: emailForDelete });
    createdUsersIds.push(userId);
    return userId;
  }
  const promiseForUserId = createUser();

  it("NO ID", async () => {
    const response = await app.inject({
      method: "DELETE",
      url: "/users/",
    });
    expect(response.statusCode).toBe(422);
    //expect(response.json()).toMatchSnapshot();
  });
  it("BAD ID", async () => {
    const response = await app.inject({
      method: "DELETE",
      url: "/users/d290f1ee-6c54-4b01-90e6-d701748f0851---",
    });
    expect(response.statusCode).toBe(422);
    //expect(response.json()).toMatchSnapshot();
  });
  it("404", async () => {
    const response = await app.inject({
      method: "DELETE",
      url: "/users/d290f1ee-6c54-4b01-90e6-d701748f0851",
    });
    expect(response.statusCode).toBe(404);
    //expect(response.json()).toMatchSnapshot();
  });
  it("200 by UUID", async () => {
    const userId = await promiseForUserId;
    const response = await app.inject({
      method: "DELETE",
      url: `/users/${userId}`,
    });
    expect(response.statusCode).toBe(200);
    const responseJson = response.json();
    //simple ok response
    expect(responseJson).toMatchSnapshot();
    const responseCheck = await app.inject({
      method: "GET",
      url: `/users/${userId}`,
    });
    expect(responseCheck.statusCode).toBe(404);
  });
});

export function fixUuidInConents(consents: UserConsent[]) {
  for (const c of consents) {
    fixUuidIn(c, "userId");
  }
}
