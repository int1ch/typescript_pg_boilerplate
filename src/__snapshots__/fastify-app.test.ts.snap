// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DELETE USERS 200 by UUID 1`] = `
Object {
  "ok": true,
}
`;

exports[`GET USERS 200 by UUID 1`] = `
Object {
  "consents": Array [],
  "email": "fixed@email.ru",
  "id": "uuid0000-fixe-df04-test-reason6922cd",
}
`;

exports[`GET USERS 404 by UUID 1`] = `
Object {
  "error": "Not Found",
  "message": "User not found d290f1ee-6c54-4b01-90e6-d701748f0851",
  "statusCode": 404,
}
`;

exports[`GET USERS BAD ID 1`] = `
Object {
  "errors": Array [
    "params/UserId must match format \\"uuid\\"",
  ],
  "message": "A validation error: params/UserId must match format \\"uuid\\"",
  "statusCode": 422,
}
`;

exports[`POST USER CONSENT EVENT POST USER CONSENT EVENT check insert insert 1`] = `
Array [
  Object {
    "enabled": false,
    "type": "email_notification",
    "userId": "uuid0000-fixe-df04-test-reason6922cd",
  },
  Object {
    "enabled": true,
    "type": "sms_notification",
    "userId": "uuid0000-fixe-df04-test-reason6922cd",
  },
]
`;

exports[`POST USER CONSENT EVENT POST USER CONSENT EVENT check insert insert 2`] = `
Array [
  Object {
    "enabled": true,
    "type": "email_notification",
    "userId": "uuid0000-fixe-df04-test-reason6922cd",
  },
  Object {
    "enabled": false,
    "type": "sms_notification",
    "userId": "uuid0000-fixe-df04-test-reason6922cd",
  },
]
`;

exports[`POST USERS / create BAD EMAIL 1`] = `
Object {
  "errors": Array [
    "body/email must match format \\"email\\"",
  ],
  "message": "A validation error: body/email must match format \\"email\\"",
  "statusCode": 422,
}
`;

exports[`POST USERS / create CREATED BY EMAIL 1`] = `
Object {
  "consents": Array [],
  "email": "fixed@email.ru",
  "id": "uuid0000-fixe-df04-test-reason6922cd",
}
`;

exports[`POST USERS / create NO EMAIL 1`] = `
Object {
  "errors": Array [
    "body must have required property 'email'",
  ],
  "message": "A validation error: body must have required property 'email'",
  "statusCode": 422,
}
`;
